File: benchmarks/churn/ref.py
# churn reference, Python. dict; del is O(1).
import sys

n = int(sys.argv[1])
live = int(sys.argv[2])
pad = "------------------------------------------"
m = {}
for i in range(n):
    m["s%d" % i] = "%d%s" % (i, pad)
    if i >= live:
        del m["s%d" % (i - live)]
acc = 0
for j in range(n - live, n):
    acc += len(m["s%d" % j])
print("live %d" % len(m))
print("checksum %d" % acc)

File: benchmarks/churn/ref.rb
# churn reference, Ruby. Hash; delete is O(1).
n = ARGV[0].to_i
live = ARGV[1].to_i
pad = "------------------------------------------"
m = {}
n.times do |i|
  m["s#{i}"] = "#{i}#{pad}"
  m.delete("s#{i - live}") if i >= live
end
acc = 0
(n - live).upto(n - 1) { |j| acc += m["s#{j}"].bytesize }
puts "live #{m.size}"
puts "checksum #{acc}"
